Implement Resizable Scroll Bar for Animation Player

migboy1 opened the pull request:
@Chenyang341 and I changed the scroll bar in Animation Player to a resizable scrollbar with draggable handles at the edges.

The location of the handles indicate where the the view in timeline starts and ends. Pulling the handles change the zoom. When one handle is pulled the other one stays in the same value.

This gives a new option to change the zoom of the timeline, facilitating its use. This seeks to close https://github.com/godotengine/godot-proposals/issues/11546

It implements the class HResizableScrollbar and can be used to implement Horizontal Resizable Scroll Bars in other places in the future.


https://github.com/user-attachments/assets/b3fc4cf2-7509-4daf-930b-67d1695a7f9f

RedMser commented:
> Would it be enough to remove the added lines in register_scene_types.cpp and the .xml files?

Yeah I think so.

Chenyang341 committed:
Implement Resizable Scroll Bar for Animation Player

Changed the scroll bar in Animation Player to a resizable scrollbar
with handles at the edges. Pulling the handles change the zoom of
the timeline. When one handle is pulled the other one stays in
the same value.

Co-authored-by: Miguel Jin <miguel.jin@tecnico.ulisboa.pt>

Chenyang341 commented:
I was uncertain about this too. In godot the extra time visible after the max time (in this case 20 seconds) is dependent in the zoom value, so when I pull the left handle inwards the extra time decreases, therefore necessary to adjust the right bound. But if the right bound is lower than the max time it stays in the same place, it is shown in the video at around 0:16.

I didn't know if it was a good idea to change the way the max visible time was calculated so I left it as it was.

KoBeWi commented on scene/gui/resizable_scroll_bar.h:
```diff
@@ -0,0 +1,103 @@
+			ResizableScrollBar(HORIZONTAL) { set_v_size_flags(0); }
+};
+
+// Vertical Resizable ScrollBar not implemented
```
Then why does it exist?

KoBeWi commented:
The new class should be in `editor/gui`, because it's only used in editor. Likewise, the theme definitions should be moved to editor theme.

KoBeWi commented:
I tested the feature and it seems to work fine for the most part. As other already mentioned, it would be better if moving one side of the scrollbar did not move the other. The worst part of the current behavior is that the other side always moves to the right, so when you drag right side of the scrollbar it shrinks additionally, and expands when dragging left side.

https://github.com/user-attachments/assets/eb48382b-4dcc-4446-8d03-ede227bb0364

Also not sure what to do about the old zoom slider. It feels redundant, but the new scrollbar does not cover all zoom levels, so it's not like we can just remove it.

Chenyang341 commented:
I have changed it so the extra time at the end stays the same, 20% of the total time (adding 4 seconds in this case). This way one side of the scrollbar no longer moves when the other side is being pulled. 

https://github.com/user-attachments/assets/e6252cc3-0dca-444c-83eb-7510ca4a2188

Repiteo commented:
Superseded by #114289; thank you for your contribution!
